sia/src/erasure_coding.rs

```diff
@@ -0,0 +1,309 @@
+use crate::rhp::{SECTOR_SIZE, SEGMENT_SIZE};
+use reed_solomon_erasure::galois_8::ReedSolomon;
+use std::io::{self, BufReader, BufWriter, Read, Write};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("ReedSolomon error: {0}")]
+    ReedSolomon(#[from] reed_solomon_erasure::Error),
+
+    #[error("IO error: {0}")]
+    Io(#[from] io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub struct ErasureCoder {
+    encoder: ReedSolomon,
+    data_shards: usize,
+    parity_shards: usize,
+}
+
+impl ErasureCoder {
+    #[allow(dead_code)]
+    pub fn new(data_shards: usize, parity_shards: usize) -> Result<Self> {
+        Ok(ErasureCoder {
+            encoder: ReedSolomon::new(data_shards, parity_shards)?,
+            data_shards,
+            parity_shards,
+        })
+    }
+
+    /// read_encoded_shards is a convenience method that reads data from the
+    /// given reader and returns erasure coded shards in a single call.
+    #[allow(dead_code)]
+    pub fn read_encoded_shards<R: io::Read>(&mut self, r: &mut R) -> Result<Vec<Vec<u8>>> {
```

Comment:
Reminder to update this to use an async reader when the async PR is merged introducing Tokio as a dependency.